rollup/fees/rollup_fee.go

```diff
@@ -0,0 +1,153 @@
+	GetState(common.Address, common.Hash) common.Hash
+}
+
+// CalculateL1MsgFee computes the L1 portion of the fee given
+// a Message and a StateDB
+// Reference: https://github.com/ethereum-optimism/optimism/blob/develop/l2geth/rollup/fees/rollup_fee.go
+func CalculateL1MsgFee(msg Message, state StateDB) (*big.Int, error) {
+	tx := asTransaction(msg)
+	raw, err := rlpEncode(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	l1BaseFee, overhead, scalar := readGPOStorageSlots(rcfg.L1GasPriceOracleAddress, state)
+	l1Fee := CalculateL1Fee(raw, overhead, l1BaseFee, scalar)
+	return l1Fee, nil
+}
+
+// asTransaction turns a Message into a types.Transaction
+func asTransaction(msg Message) *types.Transaction {
+	if msg.To() == nil {
+		return types.NewContractCreation(
+			msg.Nonce(),
+			msg.Value(),
+			msg.Gas(),
+			msg.GasPrice(),
+			msg.Data(),
+		)
+	}
+	return types.NewTransaction(
+		msg.Nonce(),
+		*msg.To(),
+		msg.Value(),
+		msg.Gas(),
+		msg.GasPrice(),
+		msg.Data(),
+	)
+}
+
+// rlpEncode RLP encodes the transaction into bytes
+// When a signature is not included, set pad to true to
+// fill in a dummy signature full on non 0 bytes
+func rlpEncode(tx *types.Transaction) ([]byte, error) {
+	raw := new(bytes.Buffer)
+	if err := tx.EncodeRLP(raw); err != nil {
+		return nil, err
+	}
+
+	r, v, s := tx.RawSignatureValues()
+	if r.Cmp(common.Big0) != 0 || v.Cmp(common.Big0) != 0 || s.Cmp(common.Big0) != 0 {
+		return nil, errTransactionSigned
+	}
+
+	// Slice off the 0 bytes representing the signature
+	b := raw.Bytes()
+	return b[:len(b)-3], nil
+}
+
+func readGPOStorageSlots(addr common.Address, state StateDB) (*big.Int, *big.Int, *big.Float) {
+	l1BaseFee := state.GetState(addr, rcfg.L1BaseFeeSlot)
+	overhead := state.GetState(addr, rcfg.OverheadSlot)
+	scalar := state.GetState(addr, rcfg.ScalarSlot)
+	scaled := ScalePrecision(scalar.Big(), rcfg.Precision)
+	return l1BaseFee.Big(), overhead.Big(), scaled
+}
+
+// ScalePrecision will scale a value by precision
+func ScalePrecision(scalar, precision *big.Int) *big.Float {
+	fscalar := new(big.Float).SetInt(scalar)
+	fdivisor := new(big.Float).SetInt(precision)
+	// fscalar / fdivisor
+	return new(big.Float).Quo(fscalar, fdivisor)
+}
+
+// CalculateL1Fee computes the L1 fee
+func CalculateL1Fee(data []byte, overhead, l1GasPrice *big.Int, scalar *big.Float) *big.Int {
+	l1GasUsed := CalculateL1GasUsed(data, overhead)
+	l1Fee := new(big.Int).Mul(l1GasUsed, l1GasPrice)
+	return mulByFloat(l1Fee, scalar)
+}
+
+// CalculateL1GasUsed computes the L1 gas used based on the calldata and
+// constant sized overhead. The overhead can be decreased as the cost of the
+// batch submission goes down via contract optimizations. This will not overflow
+// under standard network conditions.
+func CalculateL1GasUsed(data []byte, overhead *big.Int) *big.Int {
+	zeroes, ones := zeroesAndOnes(data)
+	zeroesGas := zeroes * params.TxDataZeroGas
+	onesGas := (ones + 68) * params.TxDataNonZeroGasEIP2028
```

Comment:
What does this `+ 68` represent?

Reply:
tx header size

Reply:
I see, it's for the signature. Explained here: https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts/contracts/L2/predeploys/OVM_GasPriceOracle.sol#L139-L145